### PR TITLE
Add limitation to max concurrent reindex tasks

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit adds a limitation to the max concurrent reindexing tasks
that a single cluster can perform at once. It can be configured using
a dynamic setting.

### CODE REVIEW

**Comment by tbrooks8** [comment]
POC for the method of limiting. Currently it is just a constant, but I assume that we want this to be a cluster setting?

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexMaxConcurrentTasksTests.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        indexRandom(true, docs);
+        assertHitCount(client().prepareSearch("source").setSize(0).get(), max);
+
+        try {
+            ClusterUpdateSettingsRequest settingsUpdate = new ClusterUpdateSettingsRequest();
+            Settings settings = Settings.builder().put(TransportStartReindexTaskAction.MAX_CONCURRENT_REINDEX_TASKS.getKey(), 3).build();
+            settingsUpdate.persistentSettings(settings);
+            client().admin().cluster().updateSettings(settingsUpdate).get();
+
+            // Since we are only using validation on the data nodes and this test is inherently racy, we
```
Comment (henningandersen) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexMaxConcurrentTasksTests.java]: I think we should be able to just submit 3 good requests and then validate one (or several) failing requests afterwards? I think this is only racy in unhealthy setups, AFAIK, a cluster state update waits for all nodes to have applied the state before returning (with timeout).

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexMaxConcurrentTasksTests.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                    // Copy all the docs
+                    ReindexRequestBuilder copy = reindex().source("source").destination("dest_" + i).refresh(true);
+                    StartReindexTaskAction.Request request = new StartReindexTaskAction.Request(copy.request(), false);
+                    // Use a small batch size so we have to use more than one batch
+                    copy.source().setSize(5);
+                    client().execute(StartReindexTaskAction.INSTANCE, request).actionGet();
+                }
+            });
+        } finally {
+            ClusterUpdateSettingsRequest settingsUpdate = new ClusterUpdateSettingsRequest();
```
Comment (henningandersen) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexMaxConcurrentTasksTests.java]: You might have tried this already and if so, ignore this. But could we change this specific test-class to have `@ClusterScope(scope=TEST)` instead and then not have to reset this here?

**Review by henningandersen — COMMENTED** [review]
Thanks @tbrooks8 , I left a few minor comments, otherwise this is looking good.

**Review by henningandersen — APPROVED** [review]
LGTM, thanks @tbrooks8

**Pull request merged**